File: chess/archive/chess/piece.py
from tile import Tile


class Piece:
    def __init__(self, t, w):
        self.my_type = t
        self.white = w
        self.first_move = True
        self.just_advanced_two = False

    def move(self):
        self.first_move = False

    def set_just_advanced_two(self, ju):
        self.just_advanced_two = ju

    def get_just_advanced_two(self):
        return self.just_advanced_two

    def is_white(self):
        return self.white

    def is_first_move(self):
        return self.first_move

    def is_empty(self):
        return self.my_type == 0

    def is_pawn(self):
        return self.my_type == 1

    def is_knight(self):
        return self.my_type == 2

    def is_bishop(self):
        return self.my_type == 3

    def is_rook(self):
        return self.my_type == 4

    def is_queen(self):
        return self.my_type == 5

    def is_king(self):
        return self.my_type == 6

    def to_string(self):
        path = "Images/"
        if self.is_white():
            color = "White"
        else:
            color = "Black"
        if self.my_type == 0:
            return ""
        elif self.my_type == 1:
            return path + color + "Pawn.png"
        elif self.my_type == 2:
            return path + color + "Knight.png"
        elif self.my_type == 3:
            return path + color + "Bishop.png"
        elif self.my_type == 4:
            return path + color + "Rook.png"
        elif self.my_type == 5:
            return path + color + "Queen.png"
        elif self.my_type == 6:
            return path + color + "King.png"

    def pawn_possible_moves(self, piece_layout, row, col):
        possible_moves = []

        if self.is_white():
            one = -1
            two = -2
            pawn_first_move = row == 6
        else:
            one = 1
            two = 2
            pawn_first_move = row == 1

        if pawn_first_move:
            if piece_layout[row + two][col].is_empty() and piece_layout[row + one][col].is_empty():
                possible_moves.append(Tile(row + two, col))
        if row != 0 and row != 7:
            if piece_layout[row + one][col].is_empty():
                possible_moves.append(Tile(row + one, col))
            if col != 0:
                diagonal_piece = piece_layout[row + one][col - 1]
                if diagonal_piece.is_white() != self.is_white() and not diagonal_piece.is_empty():
                    possible_moves.append(Tile(row + one, col - 1))
            if col != 7:
                diagonal_piece = piece_layout[row + one][col + 1]
                if diagonal_piece.is_white() != self.is_white() and not diagonal_piece.is_empty():
                    possible_moves.append(Tile(row + one, col + 1))

        # En passant
        if (self.is_white() and row == 3) or ((not self.is_white()) and row == 4):
            for k in range(-1, 2, 2):  # for(int k = -1; k <= 1; k += 2)
                if 0 <= col + k <= 7:
                    adj_piece = piece_layout[row][col + k]
                    if piece_layout[row + one][col + k].is_empty() \
                            and adj_piece.is_pawn() and adj_piece.get_just_advanced_two() and not adj_piece.is_white():
                        possible_moves.append(Tile(row + one, col + k))

        return possible_moves

    @staticmethod
    def knight_possible_moves(row, col):
        possible_moves = [Tile(row + 2, col + 1),
                          Tile(row + 2, col - 1),
                          Tile(row - 2, col + 1),
                          Tile(row - 2, col - 1),
                          Tile(row + 1, col + 2),
                          Tile(row + 1, col - 2),
                          Tile(row - 1, col + 2),
                          Tile(row - 1, col - 2)]

        return possible_moves

    def bishop_possible_moves(self, piece_layout, row, col):
        possible_moves = []

        # diagonal down right
        r = row + 1
        c = col + 1
        while r < 8 and c < 8:
            if piece_layout[r][c].is_empty():
                possible_moves.append(Tile(r, c))
            else:
                if not piece_layout[r][c].is_white() == self.is_white():
                    possible_moves.append(Tile(r, c))
                break
            r += 1
            c += 1

        # diagonal up left
        r = row - 1
        c = col - 1
        while r >= 0 and c >= 0:
            if piece_layout[r][c].is_empty():
                possible_moves.append(Tile(r, c))
            else:
                if not piece_layout[r][c].is_white() == self.is_white():
                    possible_moves.append(Tile(r, c))
                break
            r -= 1
            c -= 1

        # diagonal down left
        r = row + 1
        c = col - 1
        while r < 8 and c >= 0:
            if piece_layout[r][c].is_empty():
                possible_moves.append(Tile(r, c))
            else:
                if not piece_layout[r][c].is_white() == self.is_white():
                    possible_moves.append(Tile(r, c))
                break
            r += 1
            c -= 1

        # diagonal up right
        r = row - 1
        c = col + 1
        while r >= 0 and c < 8:
            if piece_layout[r][c].is_empty():
                possible_moves.append(Tile(r, c))
            else:
                if not piece_layout[r][c].is_white() == self.is_white():
                    possible_moves.append(Tile(r, c))
                break
            r -= 1
            c += 1

        return possible_moves

    def rook_possible_moves(self, piece_layout, row, col):
        possible_moves = []

        # column below
        r = row + 1
        while r < 8:
            if piece_layout[r][col].is_empty():
                possible_moves.append(Tile(r, col))
            else:
                if piece_layout[r][col].is_white() != self.is_white():
                    possible_moves.append(Tile(r, col))
                break
            r += 1

        # column above
        r = row - 1
        while r >= 0:
            if piece_layout[r][col].is_empty():
                possible_moves.append(Tile(r, col))
            else:
                if piece_layout[r][col].is_white() != self.is_white():
                    possible_moves.append(Tile(r, col))
                break
            r -= 1

        # row right
        c = col + 1
        while c < 8:
            if piece_layout[row][c].is_empty():
                possible_moves.append(Tile(row, c))
            else:
                if piece_layout[row][c].is_white() != self.is_white():
                    possible_moves.append(Tile(row, c))
                break
            c += 1

        # row left
        c = col - 1
        while c >= 0:
            if piece_layout[row][c].is_empty():
                possible_moves.append(Tile(row, c))
            else:
                if piece_layout[row][c].is_white() != self.is_white():
                    possible_moves.append(Tile(row, c))
                break
            c -= 1

        return possible_moves

    def queen_possible_moves(self, piece_layout, row, col):
        possible_moves = self.bishop_possible_moves(piece_layout, row, col) \
                         + self.rook_possible_moves(piece_layout, row, col)

        return possible_moves

    @staticmethod
    def king_possible_moves(row, col):
        possible_moves = [Tile(row + 1, col + 1),
                          Tile(row + 1, col - 1),
                          Tile(row - 1, col + 1),
                          Tile(row - 1, col - 1),
                          Tile(row + 1, col),
                          Tile(row - 1, col),
                          Tile(row, col + 1),
                          Tile(row, col - 1)]

        return possible_moves

    def castle(self, possible_moves, piece_layout, king_pos):

        pr = king_pos.get_row()
        pc = king_pos.get_col()

        if not self.in_check(piece_layout, Tile(pr, pc)):
            if self.is_king() and piece_layout[pr][pc].is_first_move():
                if piece_layout[pr][0].is_rook() and piece_layout[pr][0].is_first_move():
                    blocking = False
                    for c in range(1, 4):
                        if self.in_check(piece_layout, Tile(pr, c)) or not piece_layout[pr][c].is_empty():
                            blocking = True
                    if not blocking:
                        possible_moves.append(Tile(pr, 2))

                if piece_layout[pr][7].is_rook() and piece_layout[pr][7].is_first_move():
                    blocking = False
                    for c in range(5, 7):
                        if self.in_check(piece_layout, Tile(pr, c)) or (not piece_layout[pr][c].is_empty()):
                            blocking = True
                        if not blocking:
                            possible_moves.append(Tile(pr, 6))

    def remove_invalid_moves_first(self, piece_layout, possible_moves):
        k = 0
        while k < len(possible_moves):
            r = possible_moves[k].get_row()
            c = possible_moves[k].get_col()
            if r > 7 or r < 0 or c > 7 or c < 0 or\
                    ((not piece_layout[r][c].is_empty()) and piece_layout[r][c].is_white() == self.is_white()):
                possible_moves.remove(possible_moves[k])
                k -= 1
            k += 1

    def possible_moves(self, piece_layout, piece_pos):
        row = piece_pos.get_row()
        col = piece_pos.get_col()
        moving_piece = piece_layout[row][col]

        if moving_piece.is_empty():
            return None
        elif moving_piece.is_pawn():
            possible_moves = self.pawn_possible_moves(piece_layout, row, col)
        elif moving_piece.is_knight():
            possible_moves = self.knight_possible_moves(row, col)
        elif moving_piece.is_bishop():
            possible_moves = self.bishop_possible_moves(piece_layout, row, col)
        elif moving_piece.is_rook():
            possible_moves = self.rook_possible_moves(piece_layout, row, col)
        elif moving_piece.is_queen():
            possible_moves = self.queen_possible_moves(piece_layout, row, col)
        else:  # moving_piece.is_king():
            possible_moves = self.king_possible_moves(row, col)

        self.remove_invalid_moves_first(piece_layout, possible_moves)

        return possible_moves

    def in_check(self, piece_layout, king_pos):
        row = king_pos.get_row()
        col = king_pos.get_col()
        for r in range(8):
            for c in range(8):
                current_piece = piece_layout[r][c]
                if (not current_piece.is_empty()) and current_piece.is_white() != self.is_white():
                    possible_moves = current_piece.possible_moves(piece_layout, Tile(r, c))
                    k = 0
                    while k < len(possible_moves):
                        if possible_moves[k].get_row() == row and possible_moves[k].get_col() == col:
                            return True
                        k += 1
        return False

    def remove_invalid_moves_second(self, possible_moves, piece_layout, piece_pos, king_pos):
        pr = piece_pos.get_row()
        pc = piece_pos.get_col()

        # to make sure king can't move to block path of piece that puts it in check
        piece_is_king = piece_layout[pr][pc].is_king()

        if piece_is_king:
            self.castle(possible_moves, piece_layout, king_pos)

        k = 0
        while k < len(possible_moves):
            tr = possible_moves[k].get_row()
            tc = possible_moves[k].get_col()

            can_en_passant = piece_layout[pr][pc].is_pawn() and tc != pc and piece_layout[tr][tc].is_empty()
            en_passant_piece = None
            if can_en_passant:
                en_passant_piece = piece_layout[pr][tc]
                piece_layout[pr][tc] = Piece(0, True)

            piece_at_target = piece_layout[tr][tc]
            piece_layout[tr][tc] = piece_layout[pr][pc]
            piece_layout[pr][pc] = Piece(0, True)

            if piece_is_king:
                king_pos = Tile(tr, tc)

            if self.in_check(piece_layout, king_pos):
                possible_moves.remove(possible_moves[k])
                k -= 1

            piece_layout[pr][pc] = piece_layout[tr][tc]
            piece_layout[tr][tc] = piece_at_target
            if can_en_passant:
                piece_layout[pr][tc] = en_passant_piece

            k += 1

    def checkmate(self, piece_layout, king_pos, king_color):
        no_moves = True
        for r in range(0, 8):
            for c in range(0, 8):
                if piece_layout[r][c].is_white() == king_color and (not piece_layout[r][c].is_empty()):
                    possible_moves = piece_layout[r][c].possible_moves(piece_layout, Tile(r, c))
                    piece_layout[r][c].remove_invalid_moves_second(possible_moves, piece_layout, Tile(r, c), king_pos)
                    if len(possible_moves) != len([]):
                        no_moves = False
        if self.in_check(piece_layout, king_pos) and no_moves:
            if piece_layout[king_pos.get_row()][king_pos.get_col()].is_white():
                return 1
            else:
                return 2
        elif no_moves:
            return 3

File: chess/archive/chess-java/src/Piece.java
import javax.swing.*;
import java.util.ArrayList;

public class Piece {
    private final int myType;
    private boolean isWhite, firstMove, justAdvancedTwo;

    public Piece() {
        myType = 0;
    }

    public Piece(int t, boolean w) {
        myType = t;
        isWhite = w;
        firstMove = true;
        justAdvancedTwo = false;
    }

    public void move() {
        firstMove = false;
    }

    public void setJustAdvancedTwo(boolean j) {
        justAdvancedTwo = j;
    }

    public boolean getJustAdvancedTwo() {
        return justAdvancedTwo;
    }

    public ArrayList<Tile> validMoves(Piece[][] pieceLayout, Tile position) {
        int row = position.getRow();
        int col = position.getCol();
        Piece piece = pieceLayout[row][col];
        ArrayList<Tile> possibleMoves = new ArrayList<>();

        if (piece.isPawn()) {
            int one = isWhite ? -1 : 1;
            int two = isWhite ? -2 : 2;
            boolean pawnFirstMove = isWhite ? row == 6 : row == 1;
            if (pawnFirstMove && pieceLayout[row + two][col].isEmpty() && pieceLayout[row + one][col].isEmpty())
                possibleMoves.add(new Tile(row + two, col));
            if (row != 0 && row != 7) {
                if (pieceLayout[row + one][col].isEmpty())
                    possibleMoves.add(new Tile(row + one, col));
                if (col != 0) {
                    Piece diagonal = pieceLayout[row + one][col - 1];
                    if (diagonal.isWhite() != this.isWhite && !diagonal.isEmpty())
                        possibleMoves.add(new Tile(row + one, col - 1));
                }
                if (col != 7) {
                    Piece diagonal = pieceLayout[row + one][col + 1];
                    if (diagonal.isWhite() != this.isWhite && !diagonal.isEmpty())
                        possibleMoves.add(new Tile(row + one, col + 1));
                }
            }
            if ((this.isWhite && row == 3) || (!this.isWhite && row == 4))
                for (int k = -1; k <= 1; k += 2)
                        if (col + k >= 0 && col + k <= 7) {
                            Piece pawn = pieceLayout[3][col + k];
                            if (pieceLayout[2][col + k].isEmpty() && pawn.isPawn() && pawn.getJustAdvancedTwo() && !pawn.isWhite())
                                possibleMoves.add(new Tile(2, col + k));
                            pawn = pieceLayout[4][col + k];
                            if (pieceLayout[5][col + k].isEmpty() && pawn.isPawn() && pawn.getJustAdvancedTwo() && pawn.isWhite())
                                possibleMoves.add(new Tile(5, col + k));
                        }
        }

        else if (piece.isKnight()) {
            possibleMoves.add(new Tile(row + 2, col + 1));
            possibleMoves.add(new Tile(row + 2, col - 1));
            possibleMoves.add(new Tile(row - 2, col + 1));
            possibleMoves.add(new Tile(row - 2, col - 1));
            possibleMoves.add(new Tile(row + 1, col + 2));
            possibleMoves.add(new Tile(row + 1, col - 2));
            possibleMoves.add(new Tile(row - 1, col + 2));
            possibleMoves.add(new Tile(row - 1, col - 2));
        }

        else if (piece.isBishop()) {
            //diagonal down right
            int c = col + 1;
            for (int r = row + 1; r < 8 && c < 8; r++) {
                if (pieceLayout[r][c].isEmpty())
                    possibleMoves.add(new Tile(r, c));
                else {
                    if (!pieceLayout[r][c].isWhite() == this.isWhite)
                        possibleMoves.add(new Tile(r, c));
                    break;
                }
                c++;
            }

            //diagonal up left
            c = col - 1;
            for (int r = row - 1; r >= 0 && c >= 0; r--) {
                if (pieceLayout[r][c].isEmpty())
                    possibleMoves.add(new Tile(r, c));
                else {
                    if (!pieceLayout[r][c].isWhite() == this.isWhite)
                        possibleMoves.add(new Tile(r, c));
                    break;
                }
                c--;
            }

            //diagonal down
            // left
            c = col - 1;
            for (int r = row + 1; r < 8 && c >= 0; r++) {
                if (pieceLayout[r][c].isEmpty())
                    possibleMoves.add(new Tile(r, c));
                else {
                    if (!pieceLayout[r][c].isWhite() == this.isWhite)
                        possibleMoves.add(new Tile(r, c));
                    break;
                }
                c--;
            }

            //diagonal up right
            c = col + 1;
            for (int r = row - 1; r >= 0 && c < 8; r--) {
                if (pieceLayout[r][c].isEmpty())
                    possibleMoves.add(new Tile(r, c));
                else {
                    if (!pieceLayout[r][c].isWhite() == this.isWhite)
                        possibleMoves.add(new Tile(r, c));
                    break;
                }
                c++;
            }
        }

        else if (piece.isRook()) {
            //same column below
            for (int r = row + 1; r < 8; r++) {
                if (pieceLayout[r][col].isEmpty())
                    possibleMoves.add(new Tile(r, col));
                else {
                    if (!pieceLayout[r][col].isWhite() == isWhite)
                        possibleMoves.add(new Tile(r, col));
                    break;
                }
            }

            //same column above
            for (int r = row - 1; r >= 0; r--) {
                if (pieceLayout[r][col].isEmpty())
                    possibleMoves.add(new Tile(r, col));
                else {
                    if (!pieceLayout[r][col].isWhite() == isWhite)
                        possibleMoves.add(new Tile(r, col));
                    break;
                }
            }

            //same row to the right
            for (int c = col + 1; c < 8; c++) {
                if (pieceLayout[row][c].isEmpty())
                    possibleMoves.add(new Tile(row, c));
                else {
                    if (!pieceLayout[row][c].isWhite() == isWhite)
                        possibleMoves.add(new Tile(row, c));
                    break;
                }
            }

            //same row to the left
            for (int c = col - 1; c >= 0; c--) {
                if (pieceLayout[row][c].isEmpty())
                    possibleMoves.add(new Tile(row, c));
                else {
                    if (!pieceLayout[row][c].isWhite() == isWhite)
                        possibleMoves.add(new Tile(row, c));
                    break;
                }
            }
        }

        else if (piece.isQueen()) {
            //diagonal down right
            int c = col + 1;
            for (int r = row + 1; r < 8 && c < 8; r++) {
                if (pieceLayout[r][c].isEmpty())
                    possibleMoves.add(new Tile(r, c));
                else {
                    if (!pieceLayout[r][c].isWhite() == this.isWhite)
                        possibleMoves.add(new Tile(r, c));
                    break;
                }
                c++;
            }

            //diagonal up left
            c = col - 1;
            for (int r = row - 1; r >= 0 && c >= 0; r--) {
                if (pieceLayout[r][c].isEmpty())
                    possibleMoves.add(new Tile(r, c));
                else {
                    if (!pieceLayout[r][c].isWhite() == this.isWhite)
                        possibleMoves.add(new Tile(r, c));
                    break;
                }
                c--;
            }

            //diagonal down left
            c = col - 1;
            for (int r = row + 1; r < 8 && c >= 0; r++) {
                if (pieceLayout[r][c].isEmpty())
                    possibleMoves.add(new Tile(r, c));
                else {
                    if (!pieceLayout[r][c].isWhite() == this.isWhite)
                        possibleMoves.add(new Tile(r, c));
                    break;
                }
                c--;
            }

            //diagonal up right
            c = col + 1;
            for (int r = row - 1; r >= 0 && c < 8; r--) {
                if (pieceLayout[r][c].isEmpty())
                    possibleMoves.add(new Tile(r, c));
                else {
                    if (!pieceLayout[r][c].isWhite() == this.isWhite)
                        possibleMoves.add(new Tile(r, c));
                    break;
                }
                c++;
            }

            //same column below
            for (int r = row + 1; r < 8; r++) {
                if (pieceLayout[r][col].isEmpty())
                    possibleMoves.add(new Tile(r, col));
                else {
                    if (!pieceLayout[r][col].isWhite() == isWhite)
                        possibleMoves.add(new Tile(r, col));
                    break;
                }
            }

            //same column above
            for (int r = row - 1; r >= 0; r--) {
                if (pieceLayout[r][col].isEmpty())
                    possibleMoves.add(new Tile(r, col));
                else {
                    if (!pieceLayout[r][col].isWhite() == isWhite)
                        possibleMoves.add(new Tile(r, col));
                    break;
                }
            }

            //same row to the right
            for (int co = col + 1; co < 8; co++) {
                if (pieceLayout[row][co].isEmpty())
                    possibleMoves.add(new Tile(row, co));
                else {
                    if (!pieceLayout[row][co].isWhite() == isWhite)
                        possibleMoves.add(new Tile(row, co));
                    break;
                }
            }

            //same row to the left
            for (int co = col - 1; co >= 0; co--) {
                if (pieceLayout[row][co].isEmpty())
                    possibleMoves.add(new Tile(row, co));
                else {
                    if (!pieceLayout[row][co].isWhite() == isWhite)
                        possibleMoves.add(new Tile(row, co));
                    break;
                }
            }
        }

        else if (piece.isKing()) {
            possibleMoves.add(new Tile(row + 1, col + 1));
            possibleMoves.add(new Tile(row + 1, col - 1));
            possibleMoves.add(new Tile(row - 1, col + 1));
            possibleMoves.add(new Tile(row - 1, col - 1));
            possibleMoves.add(new Tile(row + 1, col));
            possibleMoves.add(new Tile(row - 1, col));
            possibleMoves.add(new Tile(row, col + 1));
            possibleMoves.add(new Tile(row, col - 1));
        }

        //remove invalid moves
        for (int k = 0; k < possibleMoves.size(); k++) {
            int r = possibleMoves.get(k).getRow();
            int c = possibleMoves.get(k).getCol();
            if (r > 7 || r < 0 || c > 7 || c < 0 || (!pieceLayout[r][c].isEmpty() && pieceLayout[r][c].isWhite() == this.isWhite)) {
                possibleMoves.remove(k);
                k--;
            }
        }

        return possibleMoves;
    }

    public boolean inCheck(Piece[][] pieceLayout, Tile kingPos) {
        int row = kingPos.getRow();
        int col = kingPos.getCol();
        for (int r = 0; r < 8; r++) {
            for (int c = 0; c < 8; c++) {
                Piece beingChecked = pieceLayout[r][c];
                if (!beingChecked.isEmpty() && (beingChecked.isWhite() != this.isWhite)) {
                    ArrayList<Tile> validMoves = beingChecked.validMoves(pieceLayout, new Tile(r, c));
                    for (Tile validMove : validMoves)
                        if (validMove.getCol() == col && validMove.getRow() == row)
                            return true;
                }
            }
        }
        return false;
    }

    public void removeInvalidMoves(ArrayList<Tile> possibilities, Piece[][] pieceLayout, Tile movingPiece, Tile kingPos) {
        int mr = movingPiece.getRow();
        int mc = movingPiece.getCol();

        boolean isKing = pieceLayout[mr][mc].isKing();

        boolean castle = false;
        if (isKing() && pieceLayout[mr][mc].isFirstMove()) {
            if (pieceLayout[mr][0].isRook() && pieceLayout[mr][0].isFirstMove()) {
                boolean blocking = false;
                if (!inCheck(pieceLayout, new Tile(mr, mc))) {
                    if (!(pieceLayout[mr][1].isEmpty()))
                        blocking = true;
                    for (int c = 2; c < 4; c++)
                        if (inCheck(pieceLayout, new Tile(mr, c)) || !(pieceLayout[mr][c].isEmpty()))
                            blocking = true;
                    if (!blocking) {
                        possibilities.add(new Tile(mr, 2));
                        castle = true;
                    }
                }
            }
        }
        if (pieceLayout[mr][7].isRook() && pieceLayout[mr][7].isFirstMove()) {
            boolean blocking = false;
            if (!inCheck(pieceLayout, new Tile(mr, mc))) {
                for (int c = 5; c < 7; c++)
                    if (inCheck(pieceLayout, new Tile(mr, c)) || !(pieceLayout[mr][c].isEmpty()))
                        blocking = true;
                if (!blocking) {
                    possibilities.add(new Tile(mr, 6));
                    castle = true;
                }
            }
        }


        for (int k = 0; k < possibilities.size(); k++) {
            int tr = possibilities.get(k).getRow();
            int tc = possibilities.get(k).getCol();
            boolean enPassant = pieceLayout[mr][mc].isPawn() && tc != mc && pieceLayout[tr][tc].isEmpty();
            Piece enP = null;
            if (enPassant) {
                enP = pieceLayout[mr][tc];
                pieceLayout[mr][tc] = new Piece();
            }

            if (!castle) {
                Piece atTarget = pieceLayout[tr][tc];
                pieceLayout[tr][tc] = pieceLayout[mr][mc];
                pieceLayout[mr][mc] = new Piece();

                if (isKing)
                    kingPos = new Tile(tr, tc);

                boolean inCheck = inCheck(pieceLayout, kingPos);
                if (inCheck) {
                    possibilities.remove(k);
                    k--;
                }

                pieceLayout[mr][mc] = pieceLayout[tr][tc];
                pieceLayout[tr][tc] = atTarget;
                if (enPassant)
                    pieceLayout[mr][tc] = enP;
            }
        }
    }

    public void checkmate(Piece[][] pieceLayout, Tile kingPos) {
        boolean onlyZeroes = true;
        for (int r = 0; r < pieceLayout.length; r++)
            for (int c = 0; c < pieceLayout.length; c++) {
                if (pieceLayout[r][c].isWhite() == pieceLayout[kingPos.getRow()][kingPos.getCol()].isWhite()) {
                    ArrayList<Tile> validMoves = pieceLayout[r][c].validMoves(pieceLayout, new Tile(r, c));
                    pieceLayout[r][c].removeInvalidMoves(validMoves, pieceLayout, new Tile(r, c), kingPos);
                    if (validMoves.size() != 0)
                        onlyZeroes = false;
                }
            }

        if (inCheck(pieceLayout, kingPos) && onlyZeroes) {
            JOptionPane.showMessageDialog(null, (isWhite ? "Black" : "White") + " wins!");
            System.exit(0);
        } else if (onlyZeroes) {
            JOptionPane.showMessageDialog(null, "Draw by stalemate");
            System.exit(0);
        }
    }

    public boolean isWhite() {
        return isWhite;
    }

    public boolean isFirstMove() {
        return firstMove;
    }

    public boolean isEmpty() {
        int EMPTY = 0;
        return myType == EMPTY;
    }

    public boolean isPawn() {
        int PAWN = 6;
        return myType == PAWN;
    }

    public boolean isKnight() {
        int KNIGHT = 5;
        return myType == KNIGHT;
    }

    public boolean isBishop() {
        int BISHOP = 4;
        return myType == BISHOP;
    }

    public boolean isRook() {
        int ROOK = 3;
        return myType == ROOK;
    }

    public boolean isQueen() {
        int QUEEN = 2;
        return myType == QUEEN;
    }

    public boolean isKing() {
        int KING = 1;
        return myType == KING;
    }

    public String toString() {
        if (isPawn())
            return "Pawn";
        if (isKnight())
            return "Knight";
        if (isBishop())
            return "Bishop";
        if (isRook())
            return "Rook";
        if (isQueen())
            return "Queen";
        else
            return "King";
    }
}
